DOCS-954: Add info to customlinux board doc (#2014)
viamrobotics/docs: docs/components/board/customlinux.md

> # SMEs: Susmita

Reviewer: Not sure this is right: I think of myself as the SME with Olivia being a second choice and Susmita being my third choice.

Author: ```suggestion
# SMEs: Alan, Olivia, Susmita
```

---

The `customlinux` board model supports boards like the [Pumpkin i500](https://ologicinc.com/portfolio/mediateki500/) that run Linux operating systems and are not supported by other built-in Viam models.

> To integrate a custom Linux board into your robot:

Reviewer 2: Just noting that this is the only instance of `customlinux` not all one word and monospace, in case this wasn't intended. I can see reasons for either way though, so no action unless not intended!

Author: Indeed--it was intentional because I like to use real words on occasion :) but if folks think it should be monospace I'm fine with changing it!

Reviewer 2: Not at all -- makes sense to me!

1. First [create a pin mappings file](#create-a-pin-mappings-file).
2. Then [configure a `customlinux` board](#configure-your-board) on your robot, specifying the path to the mappings file in the board config.

## Create a pin mappings file

{{% alert title="Caution" color="caution" %}}

While some lines on a board are attached to GPIO pins, some lines are attached to other board hardware.
It is important to carefully determine your `line_number` values.

> Randomly entering numbers may result in hardware damage.

Reviewer: :+1: This is an excellent caution! Nicely done.

{{% /alert %}}

The pin mappings file describes the location of each GPIO pin on the board so that `viam-server` can access the pins correctly.

> On your `customlinux` board, create a file in your <file>/home/root</file> directory called <file>board.json</file> with your board's pin mappings.

Reviewer: This is fine as an example, but the file can go anywhere else, too. You could put it in `/tmp` or `/opt` or `/etc` or `/home/nonroot_user` or anything. You can also name it anything you want, such as `mediatek_pumpkin.json` or `backyard_weeder_v2.json`.

Author: ```suggestion
On your `customlinux` board, create a JSON file in the directory of your choice with your board's pin mappings.
```

> Use the template and example below to create a JSON file with a single key, `"pins"`, whose value is a list of objects that each represent a pin on the board.

Author: ```suggestion
Use the template and example below to write the JSON file with a single key, `"pins"`, whose value is a list of objects that each represent a pin on the board.
```

{{% /tab %}}
{{< /tabs >}}

The following parameters are available for each pin object:

<!-- prettier-ignore -->
| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `name` | string | **Required** | The name of the pin. This can be anything you want but it is convenient to use the physical board pin number. <br> Example: `"3"`. |
| `device_name` | string | **Required** | The name of the device in <file>/dev</file> that this pin is attached to. Multiple pins may be attached to the same GPIO chip.  See [GPIO info tips](#tips-for-finding-gpio-information) below. <br> Example: `"gpiochip0"`.
| `line_number` | integer | **Required** | The line on the chip that is attached to this pin. See [GPIO info tips](#tips-for-finding-gpio-information) below. <br> Example: `81`. |
| `pwm_chip_sysfs_dir` | string | Optional | Uniquely specifies which PWM device within [sysfs](https://en.wikipedia.org/wiki/Sysfs) this pin is connected to. See [PWM info tips](#tips-for-finding-pwm-information) below. <br> Example: `3290000.pwm`. |

> | `pwm_id` | integer | Optional | The line number on the PWM chip. See [PWM info tips](#tips-for-finding-pwm-information) below. <br> Example: `-1`. |

Author: ```suggestion
| `pwm_id` | integer | Optional | The line number on the PWM chip. See [PWM info tips](#tips-for-finding-pwm-information) below. <br> Example: `0`. |
```

> | `pwm_id` | integer | Optional | The power management ID of the pin. See [PWM info tips](#tips-for-finding-pwm-information) below. <br> Example: `-1`. |

Reviewer: Not for power management. It's the line number on the PWM (pulse width modulated) device, similar to the line number on the GPIO device from 2 lines earlier.

Author: ```suggestion
| `pwm_id` | integer | Optional | The line number on the PWM chip. See [PWM info tips](#tips-for-finding-pwm-information) below. <br> Example: `-1`. |
```

Reviewer: I got hung up on the "power management" part, and didn't notice that the example is negative. We've been using -1 for nonexistent PWM lines, but for all the real ones, it'll be non-negative. Make the example 0 or 1.

{{% alert title="Tip" color="tip" %}}

`pwm_chip_sysfs_dir` and `pwm_id` only apply to pins with hardware PWM supported and enabled.
If your board supports hardware PWM, you will need to enable it if it is not enabled by default.
This process depends on your specific board.

{{% /alert %}}

{{% alert title="Info" color="info" %}}

The current version of `viam-server` creates PWM functionality with software.

> The implementation of hardware-based PWM for custom Linux boards is planned for release in the future, so we recommend that you add PWM information to your board now so that you do not need to update your config later.

Reviewer: :+1: Again, this is phrased very well! It doesn't draw attention to our software missing this feature, yet it does point that out clearly. Thanks! Sorry again this isn't reliably working already.

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
ls -l /sys/class/pwm
```

Here is example output from this command on a Jetson Orin AGX:

```sh {id="terminal-prompt" class="command-line" data-output="1-10"}
total 0
lrwxrwxrwx 1 root root 0 Sep  8  2022 pwmchip0 -> ../../devices/platform/3280000.pwm/pwm/pwmchip0
lrwxrwxrwx 1 root root 0 Sep  8  2022 pwmchip1 -> ../../devices/platform/32a0000.pwm/pwm/pwmchip1
lrwxrwxrwx 1 root root 0 Sep  8  2022 pwmchip2 -> ../../devices/platform/32c0000.pwm/pwm/pwmchip2
lrwxrwxrwx 1 root root 0 Sep  8  2022 pwmchip3 -> ../../devices/platform/32f0000.pwm/pwm/pwmchip3
lrwxrwxrwx 1 root root 0 Sep  8  2022 pwmchip4 -> ../../devices/platform/39c0000.tachometer/pwm/pwmchip4
```

Based on this example output, the values to use for `pwm_chip_sysfs_dir` are `328000.pwm`, `32a0000.pwm`, and so on.

Each of these directories contains a file named <file>npwm</file> containing a number.
The number in each file is the number of lines on the chip.
The `pwm_id` value will be between `0` and one less than the number of lines.

> For example, if the <file>npwm</file> contains `"4"`, then the valid `pwm_id` values are `0`, `1`, `2`, and `3`.

Reviewer: :+1: Thanks for including the example! That makes it clear that it's between 0 and one less than the npwm value, inclusive.

Author: Thank you for providing this example on the ticket!!

sudo gpioinfo
```

Here is example output from the `sudo gpioinfo` command on an Odroid C4:

```sh {id="terminal-prompt" class="command-line" data-output="1-10"}
...
  line  62:      unnamed       unused   input  active-high
  line  63:      unnamed       unused   input  active-high
  line  64:     "PIN_27"       unused   input  active-high
  line  65:     "PIN_28"       unused   input  active-high
  line  66:     "PIN_16"       unused   input  active-high
  line  67:     "PIN_18"       unused   input  active-high
  line  68:     "PIN_22"       unused   input  active-high
...
```

In this example, the human-readable names such as `"PIN_28"` indicate which physical board pin each line is attached to.
However, these names are not standardized.
Some boards have pin names like `"PH.03"`.
In either case, you need to look at the data sheet for your board and determine how the pin names map to the hardware.

### Tips for finding PWM information

Run the following command and look for unique strings within each [symlink](https://en.wikipedia.org/wiki/Symbolic_link).

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
ls -l /sys/class/pwm
```

> Here is example output from the `sudo gpioinfo` command on a Jetson Orin AGX:

Reviewer 2: ```suggestion
Here is example output from this command on a Jetson Orin AGX:
```
